File: bulletin-board-api/app/schemas/listing.py
from datetime import datetime
from enum import Enum
from uuid import UUID

import json

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.common import PaginationMeta
from app.schemas.user import UserBrief


class ListingType(str, Enum):
    SERVICE = "service"
    ITEM = "item"


class ListingMode(str, Enum):
    OFFERING = "offering"
    SEEKING = "seeking"


class Urgency(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    ASAP = "asap"


class LocationType(str, Enum):
    ON_CAMPUS = "on_campus"
    OFF_CAMPUS = "off_campus"
    REMOTE = "remote"


class ContactPreference(str, Enum):
    IN_APP = "in_app"
    EMAIL = "email"
    PHONE = "phone"


class ListingStatus(str, Enum):
    DRAFT = "draft"
    ACTIVE = "active"
    EXPIRED = "expired"
    REMOVED = "removed"
    SOLD = "sold"
    FULFILLED = "fulfilled"


class ListingCreate(BaseModel):
    type: ListingType
    listing_mode: ListingMode = ListingMode.OFFERING
    category_id: UUID
    title: str = Field(..., min_length=5, max_length=150)
    description: str = Field(..., min_length=20, max_length=5000)
    price_hint: str | None = Field(None, max_length=100)
    budget_min: float | None = Field(None, ge=0)
    budget_max: float | None = Field(None, ge=0)
    urgency: Urgency | None = None
    location_type: LocationType = LocationType.ON_CAMPUS
    location_hint: str | None = Field(None, max_length=255)
    availability: dict | None = None
    contact_preference: ContactPreference = ContactPreference.IN_APP
    contact_details: str | None = Field(None, max_length=255)
    is_regulated: bool = False
    disclaimer_accepted: bool = False

    @model_validator(mode="after")
    def validate_budget_range(self) -> "ListingCreate":
        if self.budget_min is not None and self.budget_max is not None:
            if self.budget_min > self.budget_max:
                raise ValueError("budget_min cannot exceed budget_max")
        return self


class ListingUpdate(BaseModel):
    title: str | None = Field(None, min_length=5, max_length=150)
    description: str | None = Field(None, min_length=20, max_length=5000)
    price_hint: str | None = Field(None, max_length=100)
    budget_min: float | None = Field(None, ge=0)
    budget_max: float | None = Field(None, ge=0)
    urgency: Urgency | None = None
    location_type: LocationType | None = None
    location_hint: str | None = None
    availability: dict | None = None
    contact_preference: ContactPreference | None = None
    contact_details: str | None = None
    category_id: UUID | None = None
    is_regulated: bool | None = None

    @model_validator(mode="after")
    def validate_budget_range(self) -> "ListingUpdate":
        if self.budget_min is not None and self.budget_max is not None:
            if self.budget_min > self.budget_max:
                raise ValueError("budget_min cannot exceed budget_max")
        return self


class PhotoResponse(BaseModel):
    id: UUID
    url: str
    thumbnail_url: str | None
    position: int

    model_config = {"from_attributes": True}


class CategoryBrief(BaseModel):
    id: UUID
    name: str
    slug: str

    model_config = {"from_attributes": True}


class CategoryResponse(BaseModel):
    id: UUID
    name: str
    slug: str
    listing_type: str
    description: str | None
    icon: str | None
    sort_order: int

    model_config = {"from_attributes": True}


class ListingResponse(BaseModel):
    id: UUID
    type: ListingType
    listing_mode: ListingMode
    title: str
    description: str
    price_hint: str | None
    budget_min: float | None
    budget_max: float | None
    urgency: Urgency | None
    response_count: int
    category: CategoryBrief | None
    location_type: LocationType
    location_hint: str | None
    availability: str | None
    contact_preference: ContactPreference
    is_regulated: bool
    status: ListingStatus
    view_count: int
    photos: list[PhotoResponse]
    user: UserBrief
    is_favorited: bool = False
    is_own: bool = False
    created_at: datetime
    updated_at: datetime
    expires_at: datetime

    model_config = {"from_attributes": True}

    @field_validator("availability", mode="before")
    @classmethod
    def coerce_availability(cls, v: object) -> str | None:
        if v is None:
            return None
        if isinstance(v, str):
            return v
        return json.dumps(v)


class ListingListResponse(BaseModel):
    items: list[ListingResponse]
    pagination: PaginationMeta

File: bulletin-board-frontend/src/lib/validation/listing.ts
import { z } from "zod";

export const listingCreateSchema = z
  .object({
    type: z.enum(["service", "item"], {
      required_error: "Listing type is required",
    }),
    listing_mode: z.enum(["offering", "seeking"]).default("offering"),
    category_id: z
      .string({ required_error: "Category is required" })
      .min(1, "Category is required"),
    custom_category_label: z
      .string()
      .max(80, "Custom category must be at most 80 characters")
      .optional()
      .or(z.literal("")),
    title: z
      .string({ required_error: "Title is required" })
      .min(5, "Title must be at least 5 characters")
      .max(150, "Title must be at most 150 characters"),
    description: z
      .string({ required_error: "Description is required" })
      .min(20, "Description must be at least 20 characters")
      .max(5000, "Description must be at most 5000 characters"),
    price_hint: z
      .string()
      .max(100, "Price hint must be at most 100 characters")
      .optional()
      .or(z.literal("")),
    budget_min: z.coerce
      .number()
      .min(0, "Budget cannot be negative")
      .nullish()
      .or(z.literal("")),
    budget_max: z.coerce
      .number()
      .min(0, "Budget cannot be negative")
      .nullish()
      .or(z.literal("")),
    urgency: z
      .enum(["low", "medium", "high", "asap"])
      .nullish(),
    location_type: z.enum(["on_campus", "off_campus", "remote"], {
      required_error: "Location type is required",
    }),
    location_hint: z
      .string()
      .max(200, "Location hint must be at most 200 characters")
      .optional()
      .or(z.literal("")),
    availability: z
      .string()
      .max(200, "Availability must be at most 200 characters")
      .optional()
      .or(z.literal("")),
    contact_preference: z.enum(["in_app", "email", "phone"], {
      required_error: "Contact preference is required",
    }),
    is_regulated: z.boolean().default(false),
    disclaimer_accepted: z.boolean().default(false),
  })
  .refine((data) => data.disclaimer_accepted === true, {
    message: "You must accept the terms to post a listing",
    path: ["disclaimer_accepted"],
  })
  .refine(
    (data) => {
      const min = typeof data.budget_min === "number" ? data.budget_min : null;
      const max = typeof data.budget_max === "number" ? data.budget_max : null;
      if (min != null && max != null) return min <= max;
      return true;
    },
    {
      message: "Minimum budget cannot exceed maximum budget",
      path: ["budget_max"],
    },
  );

export const listingUpdateSchema = z.object({
  title: z
    .string()
    .min(5, "Title must be at least 5 characters")
    .max(150, "Title must be at most 150 characters")
    .optional(),
  description: z
    .string()
    .min(20, "Description must be at least 20 characters")
    .max(5000, "Description must be at most 5000 characters")
    .optional(),
  category_id: z
    .string()
    .min(1, "Invalid category")
    .optional(),
  custom_category_label: z
    .string()
    .max(80, "Custom category must be at most 80 characters")
    .optional()
    .or(z.literal("")),
  price_hint: z
    .string()
    .max(100, "Price hint must be at most 100 characters")
    .nullish(),
  location_type: z
    .enum(["on_campus", "off_campus", "remote"])
    .optional(),
  location_hint: z
    .string()
    .max(200, "Location hint must be at most 200 characters")
    .nullish(),
  availability: z
    .string()
    .max(200, "Availability must be at most 200 characters")
    .nullish(),
  contact_preference: z
    .enum(["in_app", "email", "phone"])
    .optional(),
  is_regulated: z.boolean().optional(),
});

export type ListingCreateInput = z.infer<typeof listingCreateSchema>;
export type ListingUpdateInput = z.infer<typeof listingUpdateSchema>;
